rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/scan/ast/TypeInterestResolver.java

```diff
@@ -0,0 +1,118 @@
+package org.jboss.windup.rules.apps.java.scan.ast;
+
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * @author <a href="mailto:jesse.sightler@gmail.com">Jesse Sightler</a>
+ */
+public class TypeInterestResolver
+{
+    private static TypeInterestResolver defaultInstance;
+
+    private Set<TypeInterest> typeInterestSet = new LinkedHashSet<>();
+    private Map<Character, TypeInterestResolver> typeInterestMap = new HashMap<>(26);
+
+    /**
+     * Gets the default instance of the {@link TypeInterestResolver}.
+     */
+    public static TypeInterestResolver defaultInstance()
+    {
+        if (defaultInstance == null)
+        {
+            defaultInstance = new TypeInterestResolver();
+        }
+        return defaultInstance;
+    }
+
+    public TypeInterestResolver()
+    {
+    }
+
+    public void clear()
+    {
+        this.typeInterestSet.clear();
+        this.typeInterestMap.clear();
+    }
+
+    public boolean isInteresting(String packageName, String className, String methodName)
+    {
+        // package should always be at least an empty string
+        if (packageName == null)
+            packageName = "";
+
+        // class and method should be null if not specified
+        if (StringUtils.isEmpty(className))
+            className = null;
+        if (StringUtils.isEmpty(methodName))
+            methodName = null;
+
+        return isInteresting(packageName.toCharArray(), 0, className, methodName);
+    }
+
+    private boolean isInteresting(char[] packageName, int currentIndex, String className, String methodName)
+    {
+        if (currentIndex >= packageName.length)
+        {
+            return false;
+        }
+
+        Character currentCharacter = packageName[currentIndex];
+        if (currentCharacter == '.')
+            return isInteresting(packageName, currentIndex + 1, className, methodName);
+
+        TypeInterestResolver resolver = typeInterestMap.get(currentCharacter);
+        if (resolver == null)
+            return false;
+
+        if (!resolver.typeInterestSet.isEmpty())
```

Comment:
Q: Isn't this too soon? In case "com.ibm.example" comes, it will be then put inside "com.ibm.example". Then if "com.ibm.example.example2" comes, it will be registered at "com.ibm.example.example2". Now if we are searching for "com.ibm.example.example2" at the point we will get to "com.ibm.example" we will see that typeInterestSet is not empty and will search there. However this is not the right place to search for it, right?